### ISSUE TITLE
[csharp] Circular dependency prevents ApiClient instantiation

### ISSUE THREAD

**Issue by jimschubert** [swagger-api/swagger-codegen#1893]
Generated csharp-client from current http://editor.swagger.io for the petstore-simple example.

Demo of the problem here: https://github.com/jimschubert/bug-swagger-codegen-cscirc

```
git clone https://github.com/jimschubert/bug-swagger-codegen-cscirc
cd bug-swagger-codegen-cscirc
git checkout 58400545d5 # build & run -> broken
```

Build and run this commit and you'll receive the following exception:

```
System.TypeInitializationException was unhandled
  HResult=-2146233036
  Message=The type initializer for 'IO.Swagger.Client.ApiClient' threw an exception.
  Source=PetStoreSimple
  TypeName=IO.Swagger.Client.ApiClient
  StackTrace:
       at IO.Swagger.Client.ApiClient..ctor(String basePath)
       at IO.Swagger.Api.DefaultApi..ctor(String basePath) in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\PetStoreSimple\IO\Swagger\Api\DefaultApi.cs:line 195
       at ConsoleApplication.Program.Main(String[] args) in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\ConsoleApplication\Program.cs:line 14
       at System.AppDomain._nExecuteAssembly(RuntimeAssembly assembly, String[] args)
       at System.AppDomain.ExecuteAssembly(String assemblyFile, Evidence assemblySecurity, String[] args)
       at Microsoft.VisualStudio.HostingProcess.HostProc.RunUsersAssembly()
       at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
       at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
       at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
       at System.Threading.ThreadHelper.ThreadStart()
  InnerException: 
       HResult=-2146233036
       Message=The type initializer for 'IO.Swagger.Client.Configuration' threw an exception.
       Source=PetStoreSimple
       TypeName=IO.Swagger.Client.Configuration
       StackTrace:
            at IO.Swagger.Client.ApiClient..cctor() in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\PetStoreSimple\IO\Swagger\Client\ApiClient.cs:line 64
       InnerException: 
            HResult=-2147467261
            Message=Object reference not set to an instance of an object.
            Source=PetStoreSimple
            StackTrace:
                 at IO.Swagger.Client.Configuration.set_Timeout(Int32 value) in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\PetStoreSimple\IO\Swagger\Client\Configuration.cs:line 95
                 at IO.Swagger.Client.Configuration..ctor(ApiClient apiClient, Dictionary`2 defaultHeader, String username, String password, String accessToken, Dictionary`2 apiKey, Dictionary`2 apiKeyPrefix, String tempFolderPath, String dateTimeFormat, Int32 timeout) in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\PetStoreSimple\IO\Swagger\Client\Configuration.cs:line 58
                 at IO.Swagger.Client.Configuration..cctor() in C:\Users\jrs\Desktop\bug swagger-codegen\csharp-client\ConsoleApplication\PetStoreSimple\IO\Swagger\Client\Configuration.cs:line 83
            InnerException: 
```

The problem is that `ApiClient(string)` uses `Configuration.Default`, which attempts to construct a new ApiClient (... then `Configuration.Default`). The exception is thrown on setting the Timeout property of Configuration, which doesn't perform guards against the ApiClient property not existing.

I'm able to work around this (see https://github.com/jimschubert/bug-swagger-codegen-cscirc/commit/d9e7e7e2bf723d3f88383a16d10b44be00770982), but I think it should just work.

_NOTE: There's a separate issue if you look at the commits of the demo, but it's related to the example for petstore-simple generating against and older version of the api. The exception is unrelated._

**Comment by wing328** [comment]
The test cases were fine and I didn't encounter the issue before.

Recently I did encounter issue using static methods in ApiClient outside the class itself and fixed it via https://github.com/swagger-api/swagger-codegen/pull/1884, which addresses the Timeout initialisation.

Please pull the latest master and give it a try.

Also please run unit testings in https://github.com/swagger-api/swagger-codegen/tree/master/samples/client/petstore/csharp/SwaggerClientTest to see if all the tests pass in your environment.

**Comment by jimschubert** [comment]
Checked against master@bde2bbab5625f6e245d04eb1ee0390c1694df5c3, tested in Resharper and standalone NUnit runner for 2.6.4 with .NET 4 and 4.5:

![screenshot_011616_040733_pm](https://cloud.githubusercontent.com/assets/109659/12374505/4138978a-bc6b-11e5-94b8-f45281ba1c26.jpg)

ApiClient is null here in each scenario: 
```
String httpContentType = Configuration.ApiClient.SelectHeaderContentType(httpContentTypes);
```

Looking at it again, I see that changing the ApiClient assignment in the Configuration constructor to the following works by breaking the circular construction of ApiClients and Configuration objects (and tests in the test client now pass):

```
if (apiClient == null)
    ApiClient = new ApiClient(this);
        else 
    ApiClient = apiClient;
```

I'd open a pull request, but it's such a small change it'd probably be faster for you to verify and commit.

**Comment by wing328** [comment]
Thanks for the details. I'm using Mac and Mono 4.2.1 and all tests pass locally in my environment:

<img width="1095" alt="screen shot 2016-01-17 at 10 05 26 am" src="https://cloud.githubusercontent.com/assets/934260/12375545/4710ec74-bd02-11e5-9c4a-5c784e41e737.png">

so I think the issue is related to Windows (.net).

I'll look into this more and get back to you.

**Comment by jimschubert** [comment]
Weird. They pass for me on Mac, Mono 4.2.0.

```
jim at macbert in ~/temp/swagger-codegen/samples/client/petstore/csharp/SwaggerClientTest on master* bde2bba
$ mono nunit/bin/nunit-console.exe  bin/Debug/SwaggerClientTest.dll
WARNING: The runtime version supported by this application is unavailable.
Using default runtime: v4.0.30319
NUnit-Console version 2.6.4.14350
Copyright (C) 2002-2012 Charlie Poole.
Copyright (C) 2002-2004 James W. Newkirk, Michael C. Two, Alexei A. Vorontsov.
Copyright (C) 2000-2002 Philip Craig.
All Rights Reserved.

Runtime Environment -
   OS Version: Unix 15.2.0.0
  CLR Version: 4.0.30319.17020 ( Mono 4.0 ( 4.2.0 (Stable 4.2.0.179/a224653 Fri Aug 28 13:53:46 PDT 2015) ) )

ProcessModel: Default    DomainUsage: Single
Execution Runtime: mono-4.0
.........................
Tests run: 25, Errors: 0, Failures: 0, Inconclusive: 0, Time: 3.3262019 seconds
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
```

**Comment by wing328** [comment]
Right. I think the issue is bit tricky.

**Comment by wing328** [comment]
@jimschubert I've submitted #1909 to fix the issue. Test cases passed on Windows and Mac. Please help do a test if you've time.

**Comment by jimschubert** [comment]
Your fix in #1909 works as expected for me in Mac and Windows. 

When I have a little more time, I'd be interested to disassemble the code generated by Mono to understand how it is handling `beforefieldinit` on these two types versus this same code compiled with .NET 4.x. My understanding fits with the error I received in Windows.

**Comment by wing328** [comment]
I've submitted https://github.com/swagger-api/swagger-codegen/pull/1912 to avoid storing the default ApiClient in a static field in ApiClient itself.

Moving forward, default ApiClient can be found in `Configuration.Default.ApiClient`

I hope this will make it easier for developers to find the default Configuration, ApiClient and other default settings.
